src/coreclr/src/vm/prestub.cpp

```diff
@@ -1766,112 +1766,173 @@ extern "C" MethodDesc * STDCALL PreStubGetMethodDescForCompactEntryPoint (PCODE
 #endif // defined (HAS_COMPACT_ENTRYPOINTS) && defined (TARGET_ARM)
 
 //=============================================================================
-// This function generates the real code for a method and installs it into
-// the methoddesc. Usually ***BUT NOT ALWAYS***, this function runs only once
-// per methoddesc. In addition to installing the new code, this function
-// returns a pointer to the new code for the prestub's convenience.
+// This function generates the real code when from Preemptive mode.
+// It is specifically designed to work with the NativeCallableAttribute.
 //=============================================================================
-extern "C" PCODE STDCALL PreStubWorker(TransitionBlock * pTransitionBlock, MethodDesc * pMD)
+static PCODE PreStubWorker_Preemptive(
+    _In_ TransitionBlock* pTransitionBlock,
+    _In_ MethodDesc* pMD,
+    _In_opt_ Thread* currentThread)
 {
-    PCODE pbRetVal = NULL;
+    _ASSERTE(pMD->HasNativeCallableAttribute());
 
-    BEGIN_PRESERVE_LAST_ERROR;
+    PCODE pbRetVal = NULL;
 
     STATIC_CONTRACT_THROWS;
     STATIC_CONTRACT_GC_TRIGGERS;
-    STATIC_CONTRACT_MODE_COOPERATIVE;
-    STATIC_CONTRACT_ENTRY_POINT;
+    STATIC_CONTRACT_MODE_PREEMPTIVE;
 
-    MAKE_CURRENT_THREAD_AVAILABLE();
+    // Starting from preemptive mode means the possibility exists
+    // that the thread is new to the runtime so we might have to
+    // create one.
+    if (currentThread == NULL)
+    {
+        HRESULT hr;
+        currentThread = SetupThreadNoThrow(&hr);
 
-#ifdef _DEBUG
-    Thread::ObjectRefFlush(CURRENT_THREAD);
-#endif
+        // Our attempt to create a thread failed. There is nothing
+        // more we can do except fail fast. The reverse P/Invoke isn't
+        // going to work.
+        if (currentThread == NULL)
+        {
+            TerminateProcess(GetCurrentProcess(), hr);
```

Comment:
How are people going to figure what went wrong when they hit this?

Reply:
I feel like this could use a proper failfast, since that's what it was designed for. But yeah, logging wouldn't hurt too.

Reply:
I don't know the best way to author a failfast. I was copying places in the exception code paths i know are failfast. If there is a common pattern could someone point it out?

Reply:
`EEPOLICY_HANDLE_FATAL_ERROR`? That's what managed code ends up in. And it is used in vm code too.

Reply:
Thanks for the pointer. Looks like `EEPOLICY_HANDLE_FATAL_ERROR_WITH_MESSAGE()` is the most appropriate option here.